monai/engines/trainer.py

```diff
@@ -423,3 +423,282 @@ def _iteration(
             GanKeys.GLOSS: g_loss.item(),
             GanKeys.DLOSS: d_total_loss.item(),
         }
+
+
+class AdversarialTrainer(Trainer):
```

Comment:
I took a quick look and thought this class could be merged with `GanTrainer`. Alternatively, refactor `GanTrainer` into a more general base class. It appears that they all need to train a generator and discriminator.
What do you think?
cc @ericspod @Nic-Ma

Reply:
It would be a good idea to refactor GanTrainer, I don't think it's used much so there's room for improvement.

Reply:
Thinking about how we'd do adversarial training in general, I wonder if some approach where we have separate engines for each network, with training for the discriminator triggered by a handler when an iteration for the generator completes, and the double forward pass of the discriminator is done using a specialised inferer. It's possible we could eliminate specialised trainers for adversarial training completely, or at least make something more general that can handle things like CycleGAN or other more exotic setups.

Reply:
Makes sense to refactor - I've added to the refactor list. 

I seem to be having a codeformat error, I think caused by an upgrade to black?

Reply:
Yes, have created a PR to fix it.
https://github.com/Project-MONAI/MONAI/pull/7422

Reply:
ok merged in. not sure how to deal with DCO errors for other users, caused by merging in dev?

Reply:
Can the method "Rebase the branch" list here solve the issue?
https://github.com/Project-MONAI/MONAI/pull/7406/checks?check_run_id=21013930168